dev/src/main/java/org/eclipse/codewind/intellij/ui/actions/OpenContainerShellAction.java

```diff
@@ -0,0 +1,110 @@
+import com.intellij.openapi.progress.ProgressIndicator;
+import com.intellij.openapi.progress.ProgressManager;
+import com.intellij.openapi.progress.Task;
+import com.intellij.openapi.project.Project;
+import org.eclipse.codewind.intellij.core.CodewindApplication;
+import org.eclipse.codewind.intellij.core.CoreUtil;
+import org.eclipse.codewind.intellij.core.KubeUtil;
+import org.eclipse.codewind.intellij.core.Logger;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.plugins.terminal.ShellTerminalWidget;
+import org.jetbrains.plugins.terminal.TerminalView;
+
+import java.io.IOException;
+
+import static org.eclipse.codewind.intellij.ui.messages.CodewindUIBundle.message;
+
+public class OpenContainerShellAction extends AbstractProjectDependentAction {
+    public OpenContainerShellAction() {
+        super(message("ActionOpenContainerShell"));
+    }
+
+    @Override
+    protected Task.Backgroundable getTaskToRun(CodewindApplication app, Project project) {
+        return new Task.Backgroundable(project, message("ActionOpenContainerShell")) {
+
+            @Override
+            public void run(@NotNull ProgressIndicator indicator) {
+                // The shell must be opened from the UI thread.
+                CoreUtil.invokeLater(this::run0);
+            }
+
+            public void run0() {
+                // exec bash if it's installed, else exec sh
+                String command = "sh -c \"if type bash > /dev/null; then bash; else sh; fi\"";
+
+                // Open a shell in the application container
+                String processPath = null;
+                String processArgs = null;
+                if (app.connection.isLocal()) {
+                    String envPath = CoreUtil.getEnvPath();
+                    processPath = envPath != null ? envPath + "docker" : "docker"; //$NON-NLS-1$ //$NON-NLS-2$
+                    processArgs = "exec -it " + app.getContainerId() + " " + command; //$NON-NLS-1$ //$NON-NLS-2$
+                } else {
+                    processPath = KubeUtil.getCommand();
+                    if (processPath == null) {
+                        Logger.logWarning("The container shell cannot be opened because neither of the kubectl or oc commands could be found on the path");
+                        CoreUtil.openDialog(true, message("ActionOpenContainerShellErrorTitle"), message("ActionOpenContainerShellNoKubectlMsg"));
+                        return;
+                    }
+                    processArgs = "exec -n " + app.getNamespace() + " -it " + app.getPodName() + " -- " + command; //$NON-NLS-1$ //$NON-NLS-2$ //$NON-NLS-3$
+                }
+                String cmd = processPath + " " + processArgs;
+                TerminalView terminalView = TerminalView.getInstance(myProject);
+                try {
+                    ShellTerminalWidget w = terminalView.createLocalShellWidget(null);
+                    w.executeCommand(cmd);
+                    w.requestFocusInWindow();
```

Comment:
You know why the command is printed out in the console?  The docker command isn't supposed to be run inside the container.

Reply:
Eclipse doesn't show this, but VSCode does (see screen cap).  Eclipse terminal is different from either IntelliJ or VSCode.  I don't see an easy way to clear it, and I'm not sure it's worth pursuing.

![image](https://user-images.githubusercontent.com/29612372/80531106-0fa07700-8968-11ea-8091-7d17a1ff0820.png)

Reply:
Ok sounds good.